### PR TITLE
[Bugfix] Replace non-deterministic index_add_ with scatter-then-sum in TopKWeightAndReduceNaiveBatched

### LINKED ISSUE #53430: [Bug]: Unquantized BatchedTritonExperts can change greedy top-1 across identical runs
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
OS: Ubuntu 22.04.5 LTS
Kernel: Linux 5.15.0-173-generic
Machine: x86_64
Python: 3.10.12
vLLM: 0.27.1+cu129
PyTorch: 2.13.0+cu130
CUDA used to build PyTorch: 13.0
Triton: 3.7.1
Transformers: 5.15.1
GPU: NVIDIA GeForce RTX 4090
Device capability: 8.9
Driver: 580.126.09
GPU memory: 49140 MiB
```

</details>

## Summary

On vLLM `0.27.1`, repeated calls to the unquantized `BatchedTritonExperts` path can return different BF16 tensors for exactly the same inputs. The direct `fused_experts` and modular Triton controls are bitwise stable under the same check.

In a full 24-sparse-layer replay of a pinned public Qwen MoE model, the batched path's numerical variation changes the greedy next token across identical runs. The behavior is consistent with the vectorized CUDA `index_add_` in `TopKWeightAndReduceNaiveBatched.apply`. Temporarily restoring the previous per-expert reduction loop removes the repeat-run variation in my bounded check.

## Minimal reproduction

This reducer-only reproduction uses positive, row-normalized router weights and does not download a model:

```python
import torch
from vllm.model_executor.layers.fused_moe.topk_weight_and_reduce import (
    TopKWeightAndReduceNaiveBatched,
)

torch.manual_seed(1)
num_experts, num_tokens, hidden_size, topk = 60, 64, 1408, 4
device = "cuda"

topk_ids = torch.stack(
    [
        (torch.arange(num_tokens, device=device) + slot * 13) % num_experts
        for slot in range(topk)
    ],
    dim=1,
).to(torch.int32)
topk_weights = torch.rand(num_tokens, topk, device=device)
topk_weights /= topk_weights.sum(dim=1, keepdim=True)

counts = torch.bincount(topk_ids.flatten().long(), minlength=num_experts)
expert_output = torch.randn(
    num_experts,
    int(counts.max()),
    hidden_size,
    device=device,
    dtype=torch.bfloat16,
)

reducer = TopKWeightAndReduceNaiveBatched(rank=0)
outputs = [
    reducer.apply(None, expert_output, topk_weights, topk_ids, False).clone()
    for _ in range(20)
]
torch.cuda.synchronize()

for repeat, output in enumerate(outputs[1:], start=1):
    difference = (output.float() - outputs[0].float()).abs()
    print(
        repeat,
        torch.equal(output, outputs[0]),
        difference.max().item(),
        int(torch.count_nonzero(difference).item()),
    )
```

On my machine all 19 comparisons print `False`; the maximum absolute difference is `0.015625`, with approximately 11,000-12,300 changed output elements per call.

## Model-impact reproduction

I also attached `repro_vllm_unquantized_moe_path_split.py`. It has no dependency on my test harness and uses only PyTorch, Transformers, and vLLM:

```bash
python repro_vllm_unquantized_moe_path_split.py --repeats 10 > repro.json
```

This script downloads and pins:

```text
model: Qwen/Qwen1.5-MoE-A2.7B-Chat
revision: ec052fda178e241c7c443468d2fa1db6618996be
dtype: bfloat16
prompt: Continue the Python code with exactly one token:\n```python\ndef
```

It performs two checks:

1. It captures the input to the first sparse layer once, derives the router weights/IDs and expert tensors once, and invokes each vLLM MoE path repeatedly with those exact tensors.
2. It replays all 24 sparse layers in the same pinned Transformers model, changing only the sparse expert implementation, and records the final vocabulary logits and greedy next token.

The `batched_triton_loop_finalize` path is a diagnostic monkeypatch contained in the script. It replaces only `TopKWeightAndReduceNaiveBatched.apply` with the [per-expert reduction loop from vLLM
0.26.0](https://github.com/vllm-project/vllm/blob/v0.26.0/vllm/model_executor/layers/fused_moe/topk_weight_and_reduce.py).

### Fixed first-layer inputs

For 10 calls per path, comparing calls 2-10 with call 1:

| Path | Bitwise-equal repeats | Largest observed absolute difference |
| --- | ---: | ---: |
| direct `fused_experts` | 9 / 9 | 0 |
| modular Triton experts | 9 / 9 | 0 |
| batched Triton experts | 0 / 9 | 0.00390625 |

The inputs, weights, selected expert IDs, router weights, and output buffer shape are unchanged between these calls.

### Full model replay

For 10 identical forward passes:

| Sparse expert implementation | Greedy top-1 result |
| --- | --- |
| original Transformers blocks | token `73594` (three backticks), 10 / 10 |
| direct vLLM `fused_experts` | token `35597` (`calculate`), 10 / 10 |
| modular vLLM Triton experts | token `35597` (`calculate`), 10 / 10 |
| batched vLLM Triton experts | token `73594`, 5 / 10; token `35597`, 5 / 10 |

The captured router weights are model-generated rather than synthetic. Their row sums range from approximately `0.10` to `0.93` during this replay.

This is a model replay that directly invokes vLLM's MoE kernels; it is not a claim that the default single-GPU vLLM server selects `BatchedTritonExperts` for this model.

## Localization check

The current
[`TopKWeightAndReduceNaiveBatched.apply`](https://github.com/vllm-project/vllm/blob/v0.27.1/vllm/model_executor/layers/fused_moe/topk_weight_and_reduce.py)
gathers all routed contributions and finishes with:

```python
output.index_add_(0, t_idx, gathered.to(output.dtype))
```

[PyTorch documents that CUDA `index_add_` may behave nondeterministically](https://docs.pytorch.org/docs/stable/generated/torch.Tensor.index_add_.html). The observed repeat-run variation is consistent with concurrent additions to the same token row.

As a diagnostic only, I restored the earlier per-expert loop locally. With three repeats:

- the fixed first-layer output was bitwise equal on both comparisons to the first call (`2/2`), while the unmodified batched path differed on both (`0/2`);
- the full replay produced the same greedy token on all three runs (`3/3`), while the unmodified batched path produced both candidate tokens.

This local result narrows the source neighborhood, but I have not benchmarked the old loop and am not proposing it as the final implementation.

## Expected behavior

Repeated calls to the same unquantized batched MoE path with identical inputs should produce stable outputs. In particular, an internal reduction-order race should not make a downstream greedy argmax alternate between tokens.

Bitwise agreement between different MoE kernel families is not required for this report. The failure is repeat-run nondeterminism within one path.

## Scope

- Reproduced on one RTX 4090 with unquantized BF16 inputs and vLLM `0.27.1`.
- Reproduced by explicitly constructing `BatchedTritonExperts`; no distributed EP,
  all-to-all transport, quantization, or live vLLM server is required.
- I have not measured prevalence across models, GPUs, dtypes, or batch sizes.
- I have not measured the performance cost of a deterministic replacement.

## Related issue

[`#52525`](https://github.com/vllm-project/vllm/issues/52525) reports ordering-dependent nondeterminism in a quantized Marlin path. This report is in an unquantized batched finalize path and reproduces nondeterminism across repeated calls with identical inputs.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by wuhangxian**
## Purpose

Fixes #53430.

`TopKWeightAndReduceNaiveBatched.apply` used `output.index_add_(0, t_idx, gathered)` to scatter weighted expert outputs back to their source tokens. When multiple experts route to the same token (the common case with `top_k > 1`), `torch.index_add_` on CUDA performs **atomic floating-point adds** whose order of accumulation is non-deterministic across runs. Because floating-point addition is not associative, repeated calls can differ in the low bits, and those differences can propagate far enough to change a greedy argmax token and break determinism for `temperature=0` / fixed-seed inference.

## Root cause

`torch.index_add_` on CUDA uses atomic operations when multiple source indices map to the same destination. The order of atomic adds is non-deterministic, and since BF16 has limited precision, different accumulation orders produce slightly different results — enough to flip a greedy `argmax` at near-tie logits.

## Fix

Replace the atomic scatter-add with a deterministic two-step approach:

1. **Scatter** gathered expert outputs into a `(T, topk, K)` buffer where each `(token, topk-slot)` position is written at most once (no atomics, no races).
2. **Reduce** with `ops.moe_sum` — the same kernel already used by `TopKWeightAndReduceContiguous`, which accumulates in fixed top-k order with an FP32 accumulator.

This keeps the single vectorized `nonzero`/gather path and does not restore the old per-expert Python loop. The tradeoff is one temporary `(T, topk, K)` tensor.

## Related work

#53442 is an overlapping proposal for #53430. This existing PR includes regression coverage for the deterministic reduction; it is not an independent resolution of a different issue. The proposals need coordination before merge.

## Current-head validation (2026-09-17 UTC)

Head: `4cae08f0241e8eb00244a1b99c6c7f286e46a14b`. The follow-up commit only applies the repository formatter to the regression tests; production code is unchanged from `27e4e2259b5e8267f1653c07b0bf3ac290cdc14d`.

```bash
.venv/bin/python -m pytest tests/kernels/moe/test_topk_weight_and_reduce.py -q
.venv/bin/python -m pre_commit run --files \
  vllm/model_executor/layers/fused_moe/topk_weight_and_reduce.py \
  tests/kernels/moe/test_topk_weight_and_reduce.py
```

Results: **6 GPU tests passed**; all applicable pre-commit hooks passed across both changed files. Python 3.12.13, PyTorch 2.13.0+cu130, RTX PRO 5000 72GB (SM120). Editable PR source reused vLLM 0.27.1 native artifacts. Optional DeepEP could not import in this environment; these tests exercise the naive batched reducer, not DeepEP.

### Determinism and performance tradeoff

A separate old-versus-new reducer comparison used the actual pre-PR module and PR module, 60 experts, BF16 inputs, and five `(tokens, hidden size, topk)` shapes. The new reducer was bitwise identical in **19/19 repeats for every shape**. At `(8, 1408, 4)`, the old reducer matched its first output in **0/19 repeats**, while the new reducer matched in **19/19**. Some other baseline shapes happened to repeat identically; this is not a claim that every input reproduces the old race.

| Tokens | Hidden size | Top-k | Old median wall time (µs) | New median wall time (µs) | Increase | Reduction buffer (MiB) |
| --- | --- | --- | --- | --- | --- | --- |
| 1 | 1408 | 4 | 130.6 | 163.3 | 25.1% | 0.011 |
| 8 | 1408 | 4 | 133.1 | 165.5 | 24.4% | 0.086 |
| 64 | 1408 | 4 | 136.5 | 169.5 | 24.2% | 0.688 |
| 256 | 4096 | 4 | 177.1 | 191.1 | 7.9% | 8.000 |
| 1024 | 4096 | 8 | 741.3 | 781.6 | 5.4% | 64.000 |

Method: 25 warmups per implementation, then 60 synchronized wall-clock samples each with alternating execution order; input construction is outside timing, output is preallocated, and internal wrapper allocations are included. The benchmark ran on production-code head `27e4e2259b5e8267f1653c07b0bf3ac290cdc14d`. The source import path was checked explicitly. No CUDA graph was used because the dynamic `nonzero()` path synchronizes with the host.

**These are shared-GPU measurements:** an existing inference service remained running. The observed 5.4–25.1% reducer overhead is preliminary and is not an end-to-end throughput estimate. The buffer column is the size of the new `(T, topk, K)` tensor, not total peak-memory growth; its storage lifetime overlaps existing intermediates. Full-wrapper peak incremental allocations were also measured: at the largest shape, 129.372 MiB before versus 129.434 MiB after.

The independent reference sums BF16-rounded products in CPU FP64. The new result matched that BF16 reference exactly for four shapes; the largest shape had maximum absolute error 0.00048828125, consistent with FP32 versus FP64 accumulation rounding. Numerical closeness and strict bitwise repeatability were checked separately.

This fix has a measured performance cost. Maintainer review should assess whether this determinism/latency tradeoff is acceptable for the batched path. Model-generation accuracy and serving throughput were not evaluated in this rerun, so model-level validation remains outstanding.

## Earlier validation (not rerun in this workflow)

Tested on **NVIDIA RTX PRO 5000 (SM120, Blackwell)** with the minimal reproducer from the issue:

```python
import torch
from vllm.model_executor.layers.fused_moe.topk_weight_and_reduce import (
    TopKWeightAndReduceNaiveBatched,
)

torch.manual_seed(1)
num_experts, num_tokens, hidden_size, topk = 60, 64, 1408, 4
device = "cuda"

topk_ids = torch.stack(
    [(torch.arange(num_tokens, device=device) + slot * 13) % num_experts for slot in range(topk)],
    dim=1,
).to(torch.int32)
topk_weights = torch.rand(num_tokens, topk, device=device)
topk_weights /= topk_weights.sum(dim=1, keepdim=True)
counts = torch.bincount(topk_ids.flatten().long(), minlength=num_experts)
expert_output = torch.randn(num_experts, int(counts.max()), hidden_size, device=device, dtype=torch.bfloat16)

reducer = TopKWeightAndReduceNaiveBatched(rank=0)
outputs = [reducer.apply(None, expert_output, topk_weights, topk_ids, False).clone() for _ in range(20)]
torch.cuda.synchronize()

for repeat, output in enumerate(outputs[1:], start=1):
    difference = (output.float() - outputs[0].float()).abs()
    print(repeat, torch.equal(output, outputs[0]), difference.max().item(), int(torch.count_nonzero(difference).item()))
```

**Before fix:** 19/19 repeats non-deterministic (`equal=False`, `max_diff=0.015625`, ~1000-1700 changed elements per run).

**After fix:** 19/19 repeats bitwise-identical (`equal=True`, `max_diff=0.0`, `changed_elems=0`).

### Edge case tests

| Test | Result |
|------|--------|
| Basic determinism | ✅ PASS |
| Pre-allocated output tensor | ✅ PASS |
| None vs pre-allocated output match | ✅ PASS |
| `apply_router_weight_on_input=True` | ✅ PASS |
| EP rank > 0 | ✅ PASS |
| `topk=1` (no multi-expert) | ✅ PASS |
| All tokens same expert | ✅ PASS |
| `topk=8` (large) | ✅ PASS |
| FP16 dtype | ✅ PASS |
| Output size assertion | ✅ PASS |

### Modular kernel correctness (not model-generation evaluation)

Validated via `BatchedTritonExperts` + `BatchedPrepareAndFinalize` modular kernel path (the same path exercised by `test_batched_experts_end_to_end` in `tests/kernels/moe/test_batched_moe.py`) against a torch reference across 5 shapes:

| Shape (m, n, k, e, topk) | Correctness | Determinism |
|--------------------------|-------------|-------------|
| (4, 128, 256, 4, 2) | ✅ PASS | ✅ PASS |
| (8, 256, 512, 4, 2) | ✅ PASS | ✅ PASS |
| (16, 512, 512, 8, 4) | ✅ PASS | ✅ PASS |
| (32, 256, 128, 4, 2) | ✅ PASS | ✅ PASS |
| (1, 128, 64, 4, 2) | ✅ PASS | ✅ PASS |

## Tested hardware

- NVIDIA RTX PRO 5000 × 8 (Blackwell, SM120, 72GB each)
- vLLM 0.27.1, PyTorch 2.13.0+cu130, CUDA 13.0

AI assistance (OpenAI Codex) was used for analysis, test formatting, benchmarking, and server validation.

### CODE REVIEW

**Comment by wuhangxian** [comment]
Note: #53442 by @taking-lying-flat addresses the same issue with an essentially identical approach (scatter to `(T, topk, K)` buffer + `ops.moe_sum`).

Both PRs converge on the same fix independently within ~12 minutes. Either one should be fine to merge — the key design decisions are the same:

- Keep the vectorized `nonzero`/gather path (no per-expert Python loop)
- Use a `(T, topk, K)` temporary buffer so each position is written at most once
- Reduce with `ops.moe_sum` instead of `index_add_`

Happy to defer to maintainers on which to take.

**Comment by wuhangxian** [comment]
Added a regression test in `6e73846` (`tests/kernels/moe/test_topk_weight_and_reduce.py`). Recapping the issue and the fix for reviewers:

### The bug
`TopKWeightAndReduceNaiveBatched.apply` scattered weighted per-expert outputs back to their source tokens with `output.index_add_(0, t_idx, gathered)`. When `top_k > 1`, multiple experts route to the same token, so several rows land on the same output index. On CUDA `index_add_` uses atomic floating-point adds whose **accumulation order is not fixed run-to-run**; since FP addition isn't associative, repeated calls can differ in the low bits. Under temperature=0 / fixed-seed decoding those low-bit differences can flip a greedy argmax and break reproducibility.

### The fix
Replace the atomic scatter-add with a deterministic two-step reduction: scatter each expert's contribution into unique slots, then combine with `ops.moe_sum` (a compiled CUDA kernel with an FP32 accumulator). Output becomes bitwise-identical across calls.

### Reproduce / verify
On an SM120 (RTX PRO 5000) box with the issue's minimal reproducer: repeat calls went from 0/19 bitwise-identical (old path) to 19/19 (fixed). Correctness holds against a per-expert float64 reference (max_diff = 0.0).

### Test added
`test_topk_weight_and_reduce.py` covers:
- numerical equivalence vs an independent float64 reference when multiple experts contribute to the same token (`topk ∈ {1,2,3}`) — the exact case `index_add_`'s ordering affected;
- bitwise determinism across 10 repeated calls;
- the bf16 dtype path;
- `apply_router_weight_on_input=True` skipping the topk re-weight.

`moe_sum` is a compiled CUDA kernel, so the suite is GPU-gated with `skipif(not torch.cuda.is_available())`. I confirmed red→green locally: the old `index_add_` path fails the determinism assertion; the scatter-then-`moe_sum` path passes all cases.

AI assistance was used for this change; I've reviewed every line.

**Comment by wuhangxian** [comment]
@mgoin @tlrmchlsmth — following up on the test I added above, this should be ready for a look whenever one of you has a moment (you own `fused_moe` and `tests/kernels` respectively).

The one thing blocking it is CI: as a fork PR, `pre-run-check` is red and none of the jobs run until the PR is labeled. Since the fix touches a determinism-sensitive reduction, the regression test is the part that really needs to execute on a GPU runner — locally it's red on the old `index_add_` path and green on the scatter-then-`moe_sum` path, but CI confirmation would be good. Could one of you add the `ready` label so the suite runs?

On the near-duplicate: #53442 lands the same core fix independently. I've noted above I'm happy to defer to whichever you prefer — flagging only that this PR is the one carrying the regression test (bitwise-determinism across repeated calls, float64 reference for `topk ∈ {1,2,3}`, the bf16 path, and the `apply_router_weight_on_input` branch), which seems worth keeping given the bug is a non-reproducible determinism issue. Glad to rebase or adjust either way.
